libgnucash/engine/Account.cpp

```diff
@@ -6251,6 +6251,8 @@ gnc_accounts_and_all_descendants (GList *accounts)
     for (GList *n = accounts; n; n = g_list_next (n))
     {
         auto acc = static_cast<Account*>(n->data);
+        if (accset.find (acc) != accset.end())
+            continue;
         add_descendant (acc, &accset);
```

Comment:
That skips head accounts already in the set, but doesn't to anything for descendants already in the set unless the top most account happens to be first in the list.
`gnc_account_foreach_descendant` is a simple recursor so you could do a recursion function that does the same check, or you could use the rv of `insert`, e.g.

```
static void 
maybe_add_descendants (AccountSet& set, Account* acc)
{
     auto [iter, succeed] = set.insert(acc);
     if (succeed)
         for (auto child = gnc_account_get_children(acc); child; child = c_list_next(child))
             maybe_add_descendants(set, static_cast<Account*>(child->data));
}

GList*
gnc_accounts_and_all_descendants(GList *accounts)
{
    AccountSet accset;
    for (GList *n = accounts; n; n = g_list_next(n))
         maybe_add_descendants(static_cast<Account*>(n->data));
   GList *ret = nullptr;
    std::for_each (accset.begin(), accset.end(),
                   [&ret](const auto& it) { ret = g_list_prepend (ret, it); });
    return g_list_sort (ret, (GCompareFunc) acct_fullname_cmp);
}
```